File: Add_and_Search_Word_data_Structure_Design.cpp
/*
Design a data structure that supports the following two operations:

void addWord(word)
bool search(word)
search(word) can search a literal word or a regular expression string containing only letters a-z or .. A . means it can represent any one letter.

For example:

addWord("bad")
addWord("dad")
addWord("mad")
search("pad") -> false
search("bad") -> true
search(".ad") -> true
search("b..") -> true
Note:
You may assume that all words are consist of lowercase letters a-z.

click to show hint.

You should be familiar with how a Trie works. If not, please work on this problem: Implement Trie (Prefix Tree) first.
*/


//C++
///////////////////////////////////////////////////////////////////////////////////////////////
class WordDictionary {
public:
    struct TrieNode{   //create a TrieNode struct
        vector<TrieNode*> nextNode;
        bool isword;
        TrieNode():
            nextNode(vector<TrieNode*>(26,NULL)),
            isword(false){}
    };  //do not forget ";"
    
    TrieNode *root;
    WordDictionary():root(new TrieNode){}

    // Adds a word into the data structure.
    void addWord(string word) {
        TrieNode *r = root;
        for(int i = 0; i < word.size(); i++){
            if(!r->nextNode[word[i]-'a'])
                r->nextNode[word[i]-'a'] = new TrieNode;
            r = r->nextNode[word[i]-'a'];
        }
        r->isword = true;
    }

    // Returns if the word is in the data structure. A word could
    // contain the dot character '.' to represent any one letter.
    bool search(string word) {
        return searchWord(word.c_str(), root);
    }
    
    bool searchWord(const char *ch, TrieNode *r){ //const char *ch contains all value in word
        if(!r) return false;
        if(*ch == '\0') return r->isword;   //check if it is end of the word
        if(*ch == '.'){
            for(int i = 0; i < 26; i++){  //find each of the character
                if(searchWord(ch+1, r->nextNode[i])) return true;  //this should check each of them
            }
            return false;
        }
        else
            return searchWord(ch+1, r->nextNode[*ch-'a']);
        
    }
};

// Your WordDictionary object will be instantiated and called as such:
// WordDictionary wordDictionary;
// wordDictionary.addWord("word");
// wordDictionary.search("pattern");




//Java
///////////////////////////////////////////////////////////////////////////////////////////////
public class WordDictionary {
    public class TrieNode{
        public TrieNode[] next = new TrieNode[26];
        public String Tword = "";
    }
    
    public TrieNode root = new TrieNode();

    // Adds a word into the data structure.
    public void addWord(String word) {
        TrieNode r = root;
        for(char w:word.toCharArray()){ //remember "toCharArray"
            int c = w - 'a';
            if(r.next[c] == null){
                r.next[c] = new TrieNode();
            }
            r = r.next[c];
        }
        r.Tword = word;
    }

    // Returns if the word is in the data structure. A word could
    // contain the dot character '.' to represent any one letter.
    public boolean search(String word) {
        return searchWord(word.toCharArray(), 0, root);
    }
    
    public boolean searchWord(char[] ch, int k, TrieNode r){
        if(r == null){
            return false;
        }
        
        if(k == ch.length){
            return !r.Tword.equals("");
        }
        
        if(ch[k] == '.'){
            for(int i = 0; i < 26; i++){
                if(searchWord(ch, k+1, r.next[i])){
                    return true;
                }
            }
            return false;
        }else{
            return searchWord(ch, k+1, r.next[ch[k]-'a']);
        }
    }
}

// Your WordDictionary object will be instantiated and called as such:
// WordDictionary wordDictionary = new WordDictionary();
// wordDictionary.addWord("word");
// wordDictionary.search("pattern");

File: Add_and_Search_Word-Data_structure_design.swift
/*
Design a data structure that supports adding new words and finding if a string matches any previously added string.

Implement the WordDictionary class:

WordDictionary() Initializes the object.
void addWord(word) Adds word to the data structure, it can be matched later.
bool search(word) Returns true if there is any string in the data structure that matches word or false otherwise. word may contain dots '.' where dots can be matched with any letter.
 

Example:

Input
["WordDictionary","addWord","addWord","addWord","search","search","search","search"]
[[],["bad"],["dad"],["mad"],["pad"],["bad"],[".ad"],["b.."]]
Output
[null,null,null,null,false,true,true,true]

Explanation
WordDictionary wordDictionary = new WordDictionary();
wordDictionary.addWord("bad");
wordDictionary.addWord("dad");
wordDictionary.addWord("mad");
wordDictionary.search("pad"); // return False
wordDictionary.search("bad"); // return True
wordDictionary.search(".ad"); // return True
wordDictionary.search("b.."); // return True
 

Constraints:

1 <= word.length <= 500
word in addWord consists lower-case English letters.
word in search consist of  '.' or lower-case English letters.
At most 50000 calls will be made to addWord and search.
*/

/*
Solution 1:
Trie

It takes time on searching part.
when there is a . character, we recursively check if ther is matching word in Trie
*/
class WordDictionary {
    class TrieNode {
        var isWord = false
        var child = [Character: TrieNode]()
    }
    
    var root: TrieNode

    /** Initialize your data structure here. */
    init() {
        root = TrieNode()
    }
    
    /** Adds a word into the data structure. */
    func addWord(_ word: String) {
        var cur = root
        for c in word {
            if cur.child[c] == nil {
                cur.child[c] = TrieNode()
            }
            cur = cur.child[c]!
        }
        cur.isWord = true
    }
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    func search(_ word: String) -> Bool {
        return _search(Array(word), node: root)
    }
    
    func _search(_ sub: [Character], node: TrieNode) -> Bool {
        if sub.isEmpty, node.isWord { return true }
        var cur = node 
        for i in 0..<sub.count {
            let c = sub[i]
            if c == "." {
                for _node in cur.child.keys {
                    if _search(Array(sub[(i+1)...]), node: cur.child[_node]!) {
                         return true
                    }
                }
                return false
            } else {
                if cur.child[c] == nil {
                     return false
                }
                cur = cur.child[c]!
            }
        }
        return cur.isWord
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * let obj = WordDictionary()
 * obj.addWord(word)
 * let ret_2: Bool = obj.search(word)
 */

 /*
 Solution 2:
 use dictionary to map it

 // key: word.count
 // value: existing stored words
 var dict: [Int:[String]] = [:]
 */
